lib/posix-user/group.c

```diff
@@ -0,0 +1,309 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "user.h"
+
+#include <uk/essentials.h>
+#include <uk/list.h>
+#include <uk/syscall.h>
+#include <uk/assert.h>
+
+#include <unistd.h>
+#include <grp.h>
+#include <pwd.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/types.h>
+
+/* Group members */
+static char *g_members__[] = { UK_DEFAULT_USER, NULL };
+
+/* Default group */
+static struct group g__ = {
+	.gr_name = UK_DEFAULT_GROUP,
+	.gr_passwd = UK_DEFAULT_PASS,
+	.gr_gid = UK_DEFAULT_GID,
+	.gr_mem = g_members__,
+};
+
+static __uk_tls struct group_entry {
+	struct group *group;
+
+	UK_SLIST_ENTRY(struct group_entry) entries;
+} *groups_iter;
+
+UK_SLIST_HEAD(uk_group_entry_list, struct group_entry);
+
+static struct uk_group_entry_list groups;
+
+void pu_init_groups(void)
+{
+	static struct group_entry ge;
+
+	ge.group = &g__;
+	UK_SLIST_INIT(&groups);
+	UK_SLIST_INSERT_HEAD(&groups, &ge, entries);
+}
+
+void setgrent(void)
+{
+	groups_iter = UK_SLIST_FIRST(&groups);
+}
+
+void endgrent(void)
+{
+	/* Nothing to do */
+}
+
+struct group *getgrent(void)
+{
+	struct group *grp;
+
+	if (groups_iter) {
+		grp = groups_iter->group;
+		groups_iter = UK_SLIST_NEXT(groups_iter, entries);
+	} else
+		grp = NULL;
+
+	return grp;
+}
+
+UK_SYSCALL_R_DEFINE(gid_t, getgid)
+{
+	return UK_DEFAULT_GID;
+}
+
+UK_SYSCALL_R_DEFINE(int, setgid, gid_t, gid)
+{
+	/* We allow only UK_DEFAULT_GID */
+	if (unlikely(gid != UK_DEFAULT_GID))
+		return -EINVAL;
+
+	return 0;
+}
+
+/* not a syscall */
+int issetugid(void)
+{
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(gid_t, getegid)
+{
+	return UK_DEFAULT_GID;
+}
+
+/* not a syscall */
+int setegid(gid_t egid)
+{
+	/* We allow only UK_DEFAULT_GID */
+	if (unlikely(egid != UK_DEFAULT_GID))
+		return -EINVAL;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setregid, gid_t, rgid, gid_t, egid)
+{
+	/* We allow only UK_DEFAULT_GID */
+	if (unlikely(rgid != UK_DEFAULT_GID || egid != UK_DEFAULT_GID))
+		return -EINVAL;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, getresgid, gid_t *, rgid, gid_t *, egid, gid_t *, sgid)
+{
+	if (unlikely(!rgid || !egid || !sgid))
+		return -EFAULT;
+
+	*rgid = *egid = *sgid = UK_DEFAULT_GID;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setresgid, gid_t, rgid, gid_t, egid, gid_t, sgid)
+{
+	/* We allow only UK_DEFAULT_GID */
+	if (unlikely(rgid != UK_DEFAULT_GID ||
+		     egid != UK_DEFAULT_GID ||
+		     sgid != UK_DEFAULT_GID))
+		return -EINVAL;
+
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, setfsgid, uid_t, fsgid)
+{
+	return UK_DEFAULT_GID;
+}
+
+/* not a syscall */
+int initgroups(const char *user __unused, gid_t group __unused)
+{
+	return 0;
+}
+
+UK_SYSCALL_R_DEFINE(int, getgroups, int, size, gid_t *, list)
+{
+	return 0;
```

Comment:
```suggestion
	if (size <= 0)
	    return 0;
	
	UK_ASSERT(list);
	list[0] = UK_DEFAULT_GID;
	
	return 1;
```

Should this return a list containing `UK_DEFAULT_GID`?

Reply:
Actually, `initgroups`, `getgroups`, and `setgroups` are not implemented but only stubbed. But anyways, these work on supplementary groups so this would not include the default GID, but only additional ones. Since the specification does not state if the effective group ID is returned, it is fine to just return an empty list. However, I added a commit that "implements" `setgroups` so it returns `EINVAL` is an invalid GID is attempted to be set by the user. We then assume that even if the user specified a list containing only the default (valid) GID, we deduplicate this list against the effective GID. This way the list completely collapses and it is fine to return an empty list in `getgroups`.

Reply:
Ok, that makes a lot of sense, thanks.